controllers/machinehealthcheck_controller.go

```diff
@@ -2,61 +2,504 @@ package controllers
+	metrics.ObserveMachineHealthCheckShortCircuitDisabled(mhc.Name, mhc.Namespace)
+
+	utils.SetMachineCondition(mhc, &v1beta1.Condition{
+		Type:   v1beta1.RemediationAllowedCondition,
+		Status: corev1.ConditionTrue,
+	})
+
+	remediationErrors := r.remediateAll(needRemediation, resourceManager)
+	errList = append(errList, remediationErrors...)
+	if len(errList) > 0 {
+		requeueError := utilerrors.NewAggregate(errList)
+		log.V(3).Info("Reconciling: there were errors, requeuing", "errors", requeueError)
+		return reconcile.Result{}, requeueError
+	}
 
+	if requeueIn > 0 {
+		log.V(3).Info("Reconciling: some targets might go unhealthy. Ensuring a requeue happens", "requeue in", requeueIn.String())
+		return reconcile.Result{RequeueAfter: requeueIn}, nil
+	}
+
+	log.V(3).Info("Reconciling: no more targets meet unhealthy criteria")
 	return result, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *MachineHealthCheckReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1beta1.MachineHealthCheck{}).
-		Complete(r)
+func (r *MachineHealthCheckReconciler) checkHealth(targets []resources.Target) (healthy []resources.Target, needRemediation []resources.Target, overallRequeueIn time.Duration, errList []error) {
+
+	for _, t := range targets {
+		log := r.Log.WithValues("target", t.String())
+		log.Info("Reconciling: health checking")
+		needsRemediation, requeueIn, err := r.needsRemediation(t)
+		if err != nil {
+			log.Error(err, "Reconciling: failed health checking")
+			errList = append(errList, err)
+			continue
+		}
+
+		if needsRemediation {
+			needRemediation = append(needRemediation, t)
+			continue
+		}
+
+		if requeueIn > 0 {
+			log.Info("Reconciling: is likely to go unhealthy", "next check", requeueIn)
+			nodeName := "unknown"
+			if t.Node != nil {
+				nodeName = t.Node.GetName()
+			}
+			r.Recorder.Eventf(t.Node, corev1.EventTypeNormal, utils.EventReasonDetectedUnhealthy,
+				"Machine %v has unhealthy node %v", t.Machine.GetName(), nodeName,
+			)
+			overallRequeueIn = *utils.MinRequeueDuration(&overallRequeueIn, &requeueIn)
+			continue
+		}
+
+		healthy = append(healthy, t)
+	}
+
+	return
+}
+
+func (r *MachineHealthCheckReconciler) needsRemediation(t resources.Target) (bool, time.Duration, error) {
+	now := time.Now()
+	log := r.Log.WithValues("target", t.String())
+
+	// machine has failed
+	phase := pointer.StringDeref(t.Machine.Status.Phase, "")
+	if phase == v1beta1.PhaseFailed {
+		log.Info("machine is unhealthy", "phase", phase)
+		return true, time.Duration(0), nil
+	}
+
+	// the node has not been set yet
+	if t.Node == nil {
+		nodeStartupTimeout := getNodeStartupTimeout(t.MHC)
+		if nodeStartupTimeout.Seconds() == disabledNodeStartupTimeout.Seconds() {
+			// Startup timeout is disabled so no need to go any further.
+			// No node yet to check conditions, can return early here.
+			return false, 0, nil
```

Comment:
Typo?
```suggestion
			return false, time.Duration(0), nil
```

Reply:
I don't consider this as a typo, but for consistency with other return statements I will change to explicit duration